### PR TITLE
Refactor etcd deployment

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity quality robustness
/kind technical-debt test
/priority normal
/topology seed
/platform all
/exp intermediate

**What this PR does / why we need it**:
This PR refactors the deployment for the etcd and adopts the component concept. Consequently, it eliminates the Helm chart and adds full unit tests, instead. This improves several aspects and the overall code quality.

**Which issue(s) this PR fixes**:
Part of #2754 

/cc @shreyas-s-rao @amshuman-kr 
ℹ️ Similar refactoring for etcd-druid: #3099

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: pkg/operation/botanist/etcd.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+			Prefix:               common.GenerateBackupEntryName(b.Shoot.Info.Status.TechnicalID, b.Shoot.Info.Status.UID),
+			Container:            string(secret.Data[genericactuator.DataKeyBackupBucketName]),
+			FullSnapshotSchedule: snapshotSchedule,
+		})
+	}
+
+	if err := b.Shoot.Components.ControlPlane.EtcdMain.Deploy(ctx); err != nil {
+		return err
+	}
+	return b.Shoot.Components.ControlPlane.EtcdEvents.Deploy(ctx)
```
Comment (danielfoehrKn) [inline pkg/operation/botanist/etcd.go]: Can both the deployment of etcd-main and etcd-events be parallized?

File: pkg/operation/botanist/controlplane/etcd/etcd.go [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+	if foundEtcd {
+		stsName = existingEtcd.Status.Etcd.Name
+	}
+
+	existingSts, foundSts, err := e.getExistingStatefulSet(ctx, stsName)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Separate function?!
```
Comment (danielfoehrKn) [inline pkg/operation/botanist/controlplane/etcd/etcd.go]:
```suggestion
```

File: pkg/operation/botanist/controlplane/etcd/etcd.go [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+
+func (e *etcd) getExistingStatefulSet(ctx context.Context, name string) (*appsv1.StatefulSet, bool, error) {
+	obj, found, err := e.getExisting(ctx, name, &appsv1.StatefulSet{})
+	if obj != nil {
+		return obj.(*appsv1.StatefulSet), found, err
+	}
+	return nil, found, err
+}
+
+func (e *etcd) getExisting(ctx context.Context, name string, obj runtime.Object) (runtime.Object, bool, error) {
```
Comment (danielfoehrKn) [inline pkg/operation/botanist/controlplane/etcd/etcd.go]:
```suggestion
func (e *etcd) getExistingResource(ctx context.Context, name string, obj runtime.Object) (runtime.Object, bool, error) {
```

File: pkg/operation/botanist/controlplane/etcd/waiter.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			case !utils.IsTrue(etcd.Status.Ready):
+				lastErrors = multierror.Append(lastErrors, fmt.Errorf("%s is not ready yet", etcd.Name))
+			}
+		}
+
+		if lastErrors == nil {
+			return retry.Ok()
+		}
+
+		logger.Info("Waiting until the both etcds are ready...")
```
Comment (danielfoehrKn) [inline pkg/operation/botanist/controlplane/etcd/waiter.go]:
```suggestion
		logger.Info("Waiting until both the etcds are ready...")
```

**Review by danielfoehrKn — CHANGES_REQUESTED** [review]
Looks very good, did not find much. 

Compared to the master branch, the etcd mains CRD changed in the following way:
- the field backup.store.prefix value is missing the "/etcd-main" suffix
- the field etcd.quota of 8Gi is missing

This causes the etcd pods to restart. 
Also the prefix determines to my knowledge the bath in the object store - I am not sure if it is safe to change it.

**Comment by rfranzke** [comment]
Thanks @danielfoehrKn for the review! I've addressed the comments and pushed a dedicated commit so that it's easier for you to check.

> the field backup.store.prefix value is missing the "/etcd-main" suffix

That's a nice catch, thank you a lot!

> the field etcd.quota of 8Gi is missing

Yes, actually it was wrong how it was before: https://github.com/gardener/gardener/blob/master/charts/seed-controlplane/charts/etcd/templates/etcd.yaml#L60, the `.spec.backup.quota` field does not exist, see https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/etcd_types.go#L69-L100. However, the `.spec.etcd.quota` field exists (see https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/etcd_types.go#L103-L106) and I guessed that it was intended to set this one, so I fixed it in this PR. /cc @shreyas-s-rao @amshuman-kr

**Review by wyb1 — COMMENTED** [review]
Prometheus cannot scrape etcd. The config looks fine and is scraping the right target and ports. Not sure what the issue is

image

**Comment by wyb1** [comment]
I deleted all my network policies and let it reconcile and now it works. I must have had a network policy preventing the traffic. Perhaps we should add a network policy that will specifically allow traffic from prometheus to etcd? I see that etcd doesn't have the label `networking.gardener.cloud/from-prometheus: allowed`

**Comment by rfranzke** [comment]
Thanks @wyb1, I've found the problem in the `NetworkPolicy` and fixed it (see dedicated commit for it).
In this PR I don't want to introduce changes, I just want to refactor the way how etcd is being deployed. I agree that all these network policies are not optimal yet and that we should refactor them as well at a later point in time.

**Review by danielfoehrKn — DISMISSED** [review]

**Review by wyb1 — CHANGES_REQUESTED** [review]
The etcd memory dashboard contains no data. It still contains the old job name

image

**Comment by rfranzke** [comment]
@wyb1 Well, in our chat you mentioned that the dashboards don't need to be adapted, so I didn't even look after them 😉 Anyways, I've made the changes now, please check.

**Review by wyb1 — DISMISSED** [review]
I didn't think that any of the dashboards would be affected but I was wrong 😄 
Thanks for fixing!
/lgtm

File: pkg/operation/botanist/etcd_test.go [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+				defer func() { NewEtcd = oldNewEtcd }()
+				NewEtcd = validator.NewEtcd
+
+				etcd, err := botanist.DefaultEtcd(role, class)
+				Expect(etcd).NotTo(BeNil())
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
+
+		Context("no shooted seed", func() {
```
Comment (ialidzhikov) [inline pkg/operation/botanist/etcd_test.go]: There is already a context with the same name in L100.

```suggestion
		Context("for shooted seed", func() {
```

File: pkg/operation/botanist/controlplane/etcd/monitoring_alertingrules_important_with_backup.test.yaml [hunk 6]
```diff
@@ -0,0 +1,133 @@
+rule_files:
```
Comment (ialidzhikov) [inline pkg/operation/botanist/controlplane/etcd/monitoring_alertingrules_important_with_backup.test.yaml]: You could move the `*.test.yaml` files into `testdata`? Ref https://github.com/golang/go/issues/14715
```diff
- pkg/operation/botanist/controlplane/etcd/monitoring_alertingrules_important_with_backup.test.yaml
+ pkg/operation/botanist/controlplane/etcd/testdata/monitoring_alertingrules_important_with_backup.test.yaml
```

File: pkg/operation/botanist/component/types.go [hunk 7]
```diff
@@ -22,14 +22,20 @@ type Secret struct {
 	Checksum string
 }
 
-// LoggingConfig is a structure that contains additional Fluentbit filters and parsers
-type LoggingConfig struct {
-	// Filters contains the filters for specific component
+// CentralMonitoringConfig is a structure that contains configuration for the central monitoring stack.
+type CentralMonitoringConfig struct {
+	// ScrapeConfigs are the scrape configurations for centtral Prometheus.
```
Comment (ialidzhikov) [inline pkg/operation/botanist/component/types.go]:
```suggestion
	// ScrapeConfigs are the scrape configurations for central Prometheus.
```

File: pkg/operation/botanist/controlplane/etcd/etcd.go [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+	}); err != nil {
+		return err
+	}
+
+	if e.hvpaConfig != nil && e.hvpaConfig.Enabled {
+		var (
+			hpaLabels                   = map[string]string{v1beta1constants.LabelRole: "etcd-hpa-" + e.role}
+			vpaLabels                   = map[string]string{v1beta1constants.LabelRole: "etcd-vpa-" + e.role}
+			updateModeAuto              = hvpav1alpha1.UpdateModeAuto
+			UpdateModeMaintenanceWindow = hvpav1alpha1.UpdateModeMaintenanceWindow
```
Comment (ialidzhikov) [inline pkg/operation/botanist/controlplane/etcd/etcd.go]:
```suggestion
			updateModeMaintenanceWindow = hvpav1alpha1.UpdateModeMaintenanceWindow
```

File: pkg/operation/botanist/controlplane/etcd/monitoring.go [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+		return nil, err
+	}
+
+	return []string{
+		scrapeConfigEtcd.String(),
+		scrapeConfigBackupRestore.String(),
+	}, nil
+}
+
+// AlertConfig returns the alerting configuration for AlertManager.
```
Comment (ialidzhikov) [inline pkg/operation/botanist/controlplane/etcd/monitoring.go]:
```suggestion
// AlertingRules returns the alerting configuration for AlertManager.
```

File: pkg/operation/botanist/controlplane/etcd/waiter.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+func (e *etcd) Wait(_ context.Context) error        { return nil }
+func (e *etcd) WaitCleanup(_ context.Context) error { return nil }
+
+// WaitUntilEtcdsReady waits until all etcds in the given namespace are ready.
+func WaitUntilEtcdsReady(
+	ctx context.Context,
+	c client.Client,
+	logger *logrus.Entry,
```
Comment (ialidzhikov) [inline pkg/operation/botanist/controlplane/etcd/waiter.go]:
```suggestion
	logger logrus.FieldLogger,
```

**Review by ialidzhikov — COMMENTED** [review]
Looks good to me in general. I tried some operations like creation/deletion/hibernation and it worked well. Minor nits inline.

/needs second-opinion

**Comment by rfranzke** [comment]
Thanks @ialidzhikov, done!

**Review by ialidzhikov — APPROVED** [review]
I have no other comments,

/lgtm

**Pull request merged**